### PR TITLE
Reopen inductor ut test_dense_mask_index_cpu and test_expanded_reduction_cpu

### PR DESCRIPTION
**Pull request by zhuhaozhe**
Solve issue https://github.com/pytorch/pytorch/issues/93542.
Reopen 
```
test_dense_mask_index_cpu (__main__.CpuTests)
test_expanded_reduction_cpu (__main__.CpuTests)
```

The failures for these two UTs is caused by different reduce order for "sum".

For example, below code will get difference result for difference order of sum:
```python
import torch
size = 100000
a  = torch.randn(size)
b = torch.randn(3)
for i in range(10):
    c = torch.zeros((1, ))
    order = torch.randperm(size)
    for j in range(size):
        c += (a[order[j]] * b[2])
    print(c.item())
```
output:
```
1.209582805633545
1.2103385925292969
1.2097705602645874
1.2099699974060059
1.2103685140609741
1.2103989124298096
1.210781216621399
1.210976243019104
1.2101386785507202
1.2098314762115479
```

The generated CPP code cannot guarantee a fixed reduce order for "sum" with
```
 #pragma omp for reduction(+:var)
```
or guarantee exactly same reduce order for aten::sum (which is the reference to compare in UT).

We can increase the tolerance here.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @jansel @lezcano @fdrocha

### CODE REVIEW

**Comment by jgong5** [comment]
This is the side effect of omp parallel reduction which cannot guarantee a fixed reduction order among threads. The result from the parallel reduction is not necessarily less precise than the serial one (in fact, I guess it might be more precise due to shorter accumulation chain), but the indeterminism is concerning. Does it make sense to add an option to disable omp parallel reduction in cpp backend, default disabled for good perf but users can turn it on if matching eager-mode result is desired? @jansel 

In the long run, we need to explore the deterministic parallel reduction without using omp.

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -767,16 +767,16 @@ def fn(a):
...
             z = x * y
             return z.sum((0, 1))
 
-        self.common(fn, (torch.randn(2, 197, 256), torch.randn(2, 1, 256)))
+        # Mismatched elements: 1 / 256 (0.4%)
+        # Greatest absolute difference: 2.002716064453125e-05 at index (36,) (up to 1e-05 allowed)
+        # Greatest relative difference: 4.907013266928203e-06 at index (36,) (up to 1.3e-06 allowed)
+        self.common(
+            fn, (torch.randn(2, 197, 256), torch.randn(2, 1, 256)), atol=5e-5, rtol=1e-5
+        )
```
Comment (jgong5) [inline test/inductor/test_torchinductor.py:779]: The test would run on both CPU and GPU. Can we apply the looser thresholds for CPU only? Also, is there a common set of `atol` and `rtol` that we can apply to all such cases?

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -767,16 +767,16 @@ def fn(a):
...
-            )
-
         def fn(x, y):
             z = x * y
             return z.sum((0, 1))
 
-        self.common(fn, (torch.randn(2, 197, 256), torch.randn(2, 1, 256)))
+        # Mismatched elements: 1 / 256 (0.4%)
+        # Greatest absolute difference: 2.002716064453125e-05 at index (36,) (up to 1e-05 allowed)
+        # Greatest relative difference: 4.907013266928203e-06 at index (36,) (up to 1.3e-06 allowed)
```
Comment (jgong5) [inline test/inductor/test_torchinductor.py:776]: Only adding the comments of this error message seems confusing. Maybe explaining why we are using looser thresholds here is more clear?

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Comment by jgong5** [comment]
@jansel After offline chat with @zhuhaozhe , we think that making the parallel reduction deterministic first makes more sense. No need to add an option. Meanwhile, the diff between eager and inductor might still exist due to the possible different reduction order even if the result is deterministic. @zhuhaozhe will work on the deterministic parallel reduction support.

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (jansel) [reply test/inductor/test_torchinductor.py:779]: Agreed we should loosen the thresholds on cpu only.  (You can check `self.device` on the TestCase object).

Deterministic reductions on CPU also seems like a good idea, not sure if that can be done with OMP though.

**Review by jansel — CHANGES_REQUESTED** [review]
This seems

**Comment by jgong5** [comment]
> Deterministic reductions on CPU also seems like a good idea, not sure if that can be done with OMP though.

Yes. It is possible. @zhuhaozhe is working on it: https://github.com/pytorch/pytorch/pull/88411. Need to adjust it to work with vectorized cpp kernel too.

**Comment by zhuhaozhe** [comment]
Move to https://github.com/pytorch/pytorch/pull/88411.

**Pull request closed without merging**